### ISSUE TITLE
API-47: MicroProfile Context Propagation 1.2 in OpenLiberty

### ISSUE THREAD

**Issue by njr-11** [OpenLiberty/open-liberty#8750]
Keep Open Liberty up-to-date with MicroProfile Context Propagation 1.2 as enhancements are added to the specification, and deliver under the mpContextPropagation-1.2 feature.

Upcoming feature overview:
https://ibm.box.com/s/qo7dm71wm3yb2vy3sodlvapki84h5e9w
- FTS: https://github.com/openliberty/open-liberty/issues/14718

The issue that tracks the related openliberty.io doc (existing doc is still applicable):
https://github.com/OpenLiberty/docs/issues/3562

**Comment by atosak** [comment]
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (April 1)
 
Instructions:
 - Do the actions below and mark them complete in the checklist when they are done.
 - Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **Feb 5th** 
- [ ] POC Design / WAD Review Scheduled (David Chang) or N/A.
- [ ] POC Design / WAD Reviewed (Feature Owner) or N/A.
- [ ] Complete any follow-ons from the POC Review.
- [ ] Design / WAD Approval (Alasdair Nottingham) or N/A.
- [ ] No Design / No WAD Approval (Arthur De Magalhaes - cloud / Alasdair Nottingham - server) or N/A.
- [ ] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [ ] ID Requirements identified. (Epic owner / Feature owner with ID focal point)
- [ ] Create a child task of the epic entitled "FAT Approval Test Summary".  Add and fill in the template as described here: https://github.ibm.com/was-liberty/WS-CD-Open/wiki/Feature-Review-(Feature-Test-Summary-Process)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **March 11th** 
- [ ] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE ** March 11th** 
- [ ] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **March 18th** 
- [ ] Implementation complete. (Epic owner / Feature owner) 
- [ ] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [ ] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## APPROVALS with TARGET COMPLETION DATE **March 18th to 25th** 
Prereq: You must have the Design Approved or No Design Approved label on the GitHub Epic.
- [ ] **Accessibility** - (G Scott Johnston). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [ ] **FAT Liberty SOE** - (Kevin Smith). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [ ] **Globalization** (Sam Wong - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.
- [ ] **ID** - (Kareen Deen). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [ ] **Performance** - (Jared Anderson). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [ ] **Serviceability** - (Don Bourne). Serviceability has been addressed. 
- [ ] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [ ] **SVT** - (Greg Ecock - Cloud, Brian Hanczaryk- APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.
- [ ] **Demo** - (Liberty only - Tom Evans or Chuck Bridgham). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## TARGET COMPLETION DATE **March 25th** 
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## NOT REQUIRED FOR A FEATURE
- [ ] **OL Guides** - (Yee-Kang Chang). Assessment for OL Guides is complete or N/A. 
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
 
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Related Deliverables TARGET COMPLETION DATE **April 9th** 
- [ ] Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by kwsutter** [comment]
**UFO Review** Held on Oct 23, 2020
Recording:  https://ibm.box.com/s/ssajcoe4tdr0r3lvwudtoqlsfjndwchy

**Summary:**  Good, solid review.  Some voiced concerns about the changes going into MP Config 2.0 that forced some of these updates into Context Propagation 1.1.  But, that does not affect MP Context Propagation 1.1 moving forward. They are just reacting to the direction of MP Config 2.0.

**Minutes:**
- MP Context Propagation is still a standalone component (in the MicroProfile Platform sense).  Nathan expressed an interest in becoming part of the Platform, but that's a separate topic.  MP Context Propagation already supports Jakarta EE 8.
- @tjwatson will follow up to see if the issues that @andymc12  is having with GraphQL's support of both MP 3.0 and 4.0 also apply to MP Context Propagation.  Brief conversation indicates that maybe it doesn't apply, but Tom will follow up and reach out to Nathan if necessary.
- The API Update should be referenced on page 4 of the chart deck (why are we doing this 1.1 release?).
- Much discussion on the Config 2.0 breaking change that ripples into MP Context Propagation (and possibly other MP specifications).
- @jtmulvey will follow up with stack products that use MP Context Propagation to let them know about this breaking change and see if they are affected.
- @kwsutter will follow up with @Emily-Jiang to discuss the impact of this MP Config 2.0 breaking change to all of the other MP specs and customer applications.  @njr-11 mentioned that an issue for Config to standardize on the "empty string" definition (ie. the use of "None") was already closed.
- Need to update the chart about Beta first, then GA.
- Need to update the chart to indicate that zero migration is **not** broken.  The breaking changes only come into play if the customer changes their configuration.
- Some discussion on the potential migration impacts for these Config 2.0 and Context Propagation 1.1 changes.  @cthigh will update an existing Issue that talks about migrating from MP X to MP Y (currently in "will not implement" state).  @yeekangc will create a feature against the LSP4MP project with a similar intent.  Somehow detecting application's code and/or configuration that might be affected by these changes.
- @jhanders34 noted that performance will still be executed to ensure that none of the proposed changes will affect MP Context Propagation performance.

**Comment by njr-11** [comment]
Slides have been updated per the above items.  No changes are made regarding item 2.  The MP Context Propagation 1.1 TCK has been demonstrated to run cleanly against both MP Config 2.0 (from MicroProfile 4.0) and MP Config 1.x from earlier MicroProfile versions.  Instead of an autofeature, MP Context Propagation uses dynamic import package for access to MP Config interfaces. The resulting bundle does not enforce a version range on MP Config, allowing the same bundle to be used against any MP Config version.

**Comment by njr-11** [comment]
The design call regarding item 2 was this morning and no issues were identified that would cause trouble for MP Context Propagation 1.1.  GraphQL is in a different situation because it actually depends on mpConfig and needs to tolerate across versions.   With MP Context Propagation, mpConfig is completely optional and so there is no tolerates version info for it in our features.

**Comment by njr-11** [comment]
@donbourne please add the focalApproved:serviceability label.

Serviceability Approval Comment:

WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? Have these issues been addressed in the implementation?
_We cover this in the Skill Transfer, not the WAD/UFO. But otherwise, yes, the implementation addresses the known serviceability issues._

Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).
a) What problem paths were tested and demonstrated?
_MP Config 2.0 breaking change around empty list values that impacts some of our properties for setting default context types._
b) Who did you demo to?
_Greg Watts, James Stephens, Kyle Aure_
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?
_Yes to the extent that it is possible. There is nothing we can do about the MP Config 2.0 breaking change that impacts us and other general users of MP Config. Possibly that could be something for the MP Config 2.0 feature to consider in their own serviceability review._

SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
_Brian Hanczaryk, who determined that SVT was unnecessary due to exhaustive coverage in FAT_
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?
_If SVT has any concerns with serviceability or otherwise with this feature then they will withhold their SVT approval._

Which L2 / L3 queues will handle PMRs for this feature? Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it. Ask Don Bourne if you need links or more info.
_WAS L2: CET
WAS L3: EEConcurrency, AsyncBeans_

**Comment by skasund** [comment]
L2 has requested STE slides for this feature.  The STE template can be found at the links below.  You can use either one to create the education.  

Slide Template:  https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm

Github Template:  https://pages.github.ibm.com/WASL3/site/STE/about

Please upload the completed slides to the same 'STE Archive' BOX folder or provide me the Github link.   Thanks!

**Comment by njr-11** [comment]
@skasund the STE can be found here
https://ibm.box.com/s/v4e7bx9jk7psdyb565mqz5foxsqux3ar

**Comment by skasund** [comment]
@njr-11  Thanks for the STE Slides

**Comment by chirp1** [comment]
From discussions with Nathan, no doc updates are required by the ID team. Approving.
